README.md
# oktoast

[![oktoast](https://img.shields.io/badge/OpenFlutter-OKToast-blue.svg)](https://github.com/OpenFlutter/flutter_oktoast)
[![pub package](https://img.shields.io/pub/v/oktoast.svg)](https://pub.dartlang.org/packages/oktoast)
![GitHub](https://img.shields.io/github/license/OpenFlutter/flutter_oktoast.svg)
[![GitHub stars](https://img.shields.io/github/stars/OpenFlutter/flutter_oktoast.svg?style=social&label=Stars)](https://github.com/OpenFlutter/flutter_oktoast)

A library for flutter.

A pure dart toast Library.

You can completely customize the style of toast.

[中文博客介绍](https://www.kikt.top/posts/flutter/toast/oktoast/)

## Screenshot

| Default                                                                                | Custom                                                                                 | GIF                                                                                 |
|----------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------|
| ![pic](https://raw.githubusercontent.com/CccFans/images/master/img/20181207161700.png) | ![pic](https://raw.githubusercontent.com/CccFans/images/master/img/20181207161742.png) | ![pic](https://raw.githubusercontent.com/CaiJingLong/some_asset/master/ktoast2.gif) |

## Versions

### 3.x.x

Starting from the 3.x version, OKToast provides a null-safety version,
the specific introduction of null-safety can be viewed in [dart][dart-safe] or [flutter][flutter-safe].

[flutter-safe]: https://flutter.dev/docs/null-safety
[dart-safe]: https://dart.dev/null-safety

The 2.3.2 version is the last version that does not support null-safety.

### About version 1.x

if you use OKToast 1.x, Please use the 1.x branch, and read version readme.

Proposed migration to 2.x version. The new version does not require buildContext.

And you can completely customize the style of toast, because now you can use `showToastWidget`.

## Usage

### 1. Add library to your pubspec.yaml

```shell
flutter pub add oktoast
```

### 2. Import library in dart file

```dart
import 'package:oktoast/oktoast.dart';
```

### 3. Wrap your app widget

```dart
OKToast(
  /// set toast style, optional
  child:MaterialApp()
);
```

Tips:
If you happened error like: `No MediaQuery widget found`,
you can try to use this [code](https://github.com/OpenFlutter/flutter_oktoast/issues/53#issuecomment-628431625)
to include `OKToast` to your App.

```dart
MaterialApp(
  builder: (BuildContext context, Widget? widget) {
    return OKToast(child: widget);
  },
);
```

### 4. Call method `showToast`

```dart
showToast('content');

// position and second have default value, is optional
showToastWidget(Text('hello oktoast'));
```

## Explain

There are two reasons why you need to wrap MaterialApp

1. Because this ensures that toast can be displayed in front of all other controls
2. Context can be cached so that it can be invoked anywhere without passing in context

## Properties

### OKToast params

OKToast have default style, and you also can custom style or other behavior.

|         name         |          type           |   need   |                             desc                              |
|:--------------------:|:-----------------------:|:--------:|:-------------------------------------------------------------:|
|        child         |         Widget          | required |                     Usually Material App                      |
|      textStyle       |        TextStyle        | optional |                                                               |
|        radius        |         double          | optional |                                                               |
|   backgroundColor    |          Color          | optional |                        backroundColor                         |
|       position       |      ToastPosition      | optional |                                                               |
|  dismissOtherOnShow  |          bool           | optional |    If true, other toasts will be dismissed. Default false.    |
| movingOnWindowChange |          bool           | optional | If true, when the size changes, toast is moved. Default true. |
|    textDirection     |      TextDirection      | optional |                                                               |
|     textPadding      |   EdgeInsetsGeometry    | optional |                     Outer margin of text                      |
|      textAlign       |        TextAlign        | optional |          When the text wraps, the align of the text.          |
|     handleTouch      |          bool           | optional | Default is false, if it's true, can responed use touch event. |
|   animationBuilder   | OKToastAnimationBuilder | optional |              Add animation to show / hide toast.              |
|  animationDuration   |        Duration         | optional |                  The duration of animation.                   |
|    animationCurve    |          Curve          | optional |                      Curve of animation.                      |
|       duration       |        Duration         | optional |                  Default duration of toast.                   |

### Method `showToast`

Display text on toast.

Description of params see `OKToast`.

|       name        |          type           |   need   |      desc      |
|:-----------------:|:-----------------------:|:--------:|:--------------:|
|        msg        |         String          | required | Text of toast. |
|      context      |      BuildContext       | optional |                |
|     duration      |        Duration         | optional |                |
|     position      |      ToastPosition      | optional |                |
|     textStyle     |        TextStyle        | optional |                |
|    textPadding    |   EdgeInsetsGeometry    | optional |                |
|  backgroundColor  |          Color          | optional |                |
|      radius       |         double          | optional |                |
|     onDismiss     |        Function         | optional |                |
|   textDirection   |      TextDirection      | optional |                |
| dismissOtherToast |          bool           | optional |                |
|     textAlign     |        TextAlign        | optional |                |
| animationBuilder  | OKToastAnimationBuilder | optional |                |
| animationDuration |        Duration         | optional |                |
|  animationCurve   |          Curve          | optional |                |

### Method `showToastWidget`

Display custom widgets on toast

Description of params see `showToast`.

|       name        |          type           |   need   |              desc               |
|:-----------------:|:-----------------------:|:--------:|:-------------------------------:|
|      widget       |         Widget          | required | The widget you want to display. |
|      context      |      BuildContext       | optional |                                 |
|     duration      |        Duration         | optional |                                 |
|     position      |      ToastPosition      | optional |                                 |
|     onDismiss     |        Function         | optional |                                 |
| dismissOtherToast |          bool           | optional |                                 |
|   textDirection   |      TextDirection      | optional |                                 |
|    handleTouch    |          bool           | optional |                                 |
| animationBuilder  | OKToastAnimationBuilder | optional |                                 |
| animationDuration |        Duration         | optional |                                 |
|  animationCurve   |          Curve          | optional |                                 |

### Method `dismissAllToast`

Dismiss all toast.

### Return value of `showToast` and `showToastWidget`

about return type:  
`showToast` and `showToastWidget` return type is `ToastFuture`,
The `ToastFuture` can be use to dismiss the toast.

### For all dismiss toast method

An optional parameter `showAnim` is added to control whether fading animation is required for dismiss.

The praram default value is `false`.

## Examples

```dart
import 'package:flutter/material.dart';
import 'package:oktoast/oktoast.dart'; // 1. import library

void main() => runApp( MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return OKToast(
      // 2. wrap your app with OKToast
      child:  MaterialApp(
        title: 'Flutter Demo',
        theme:  ThemeData(
          primarySwatch: Colors.blue,
        ),
        home:  MyHomePage(),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() =>  _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    _counter++;

    // 3.1 use showToast method
    showToast(
      "$_counter",
      duration: Duration(seconds: 2),
      position: ToastPosition.bottom,
      backgroundColor: Colors.black.withOpacity(0.8),
      radius: 13.0,
      textStyle: TextStyle(fontSize: 18.0),
    );

    showToast(
      "$_counter",
      duration: Duration(seconds: 2),
      position: ToastPosition.top,
      backgroundColor: Colors.black.withOpacity(0.8),
      radius: 3.0,
      textStyle: TextStyle(fontSize: 30.0),
    );

    // 3.2 use showToastWidget method to custom widget
    Widget widget = Center(
      child: ClipRRect(
        borderRadius: BorderRadius.circular(30.0),
        child: Container(
          width: 40.0,
          height: 40.0,
           color: Colors.grey.withOpacity(0.3),
          child: Icon(
            Icons.add,
            size: 30.0,
            color: Colors.green,
          ),
        ),
      ),
    );

    ToastFuture toastFuture = showToastWidget(
      widget,
      duration: Duration(seconds: 3),
      onDismiss: () {
        print("the toast dismiss"); // the method will be called on toast dismiss.
      },
    );

    // can use future
    Future.delayed(Duration(seconds: 1), () {
      toastFuture.dismiss(); // dismiss
    });

    setState(() {

    });
  }

  @override
  Widget build(BuildContext context) {
    return  Scaffold(
      appBar:  AppBar(
        title:  Text("ktoast demo"),
      ),
      body: Stack(
        children: <Widget>[
           Center(
            child: ListView(
              children: <Widget>[
                 Text(
                  'You have pushed the button this many times:',
                ),
                 Text(
                  '$_counter',
                  style: Theme.of(context).textTheme.display1,
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: RaisedButton(
                    onPressed: () {
                      Navigator.push(context,
                          MaterialPageRoute(builder: (ctx) => MyHomePage()));
                    },
                  ),
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: RaisedButton(
                    onPressed: _incrementCounter,
                    child: Text('toast'),
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}


```

## CHANGELOG

[Link](https://github.com/OpenFlutter/flutter_oktoast/blob/master/CHANGELOG.md)

## LICENSE

[Apache-2.0](https://github.com/OpenFlutter/flutter_oktoast/blob/master/LICENSE)


CHANGELOG.md
# Change log

## 3.4.0

Feat:

- Wrapper a `Material` widget when the `OKToast` widget is not wrapped with `Material`.
- The min support of flutter SDK version to 3.7.0.

## 3.3.2+1

- Fix `Theatre` constructor and get rid of `View` for compatibilities. (#103)

## 3.3.2

- Fix position offset do not work when `movingOnWindowChange` is false. (#100)

## 3.3.1

- Replace `context.debugDoingBuild` flag with `SchedulerBinding.instance.schedulerPhase`. (#97)

## 3.3.0

- Abstract `BuildContextPredicate`. (#95)
- More text fields for `showToast`. (#96)

## 3.2.0

Migrate to Flutter 3, drop supports for previous Flutter versions.

- Remove deprecated `Miui10AnimBuilder`.

## 3.1.5

- `_ToastContainer` -> `ToastContainer`.
- Add tests.

## 3.1.4

- Fix dismiss is invalid when called synchronously. (#88)
- Deprecated inner `Miui10AnimBuilder`.

## 3.1.3+1

- Quick fix for the `_throwIfNoContext`.

## 3.1.3

- Throw a more precise error if the `OKToast` widget is not wrapped correctly.
- Bind `ToastFuture` with `OverlayEntry` mounted state.

## 3.1.2

- Deliver overlay insertion into microtask.

## 3.1.1

- Expose `constraints` and `margin` for `showToast` method. (#84)

## 3.1.0

- Support `copyWith` for `ToastPosition`. (#82)
- Integrate `animationBuilder` properly. (#81)
- **BREAKING CHANGE**: `handleTouth` -> `handleTouch`.
- Code cleanup.

## 3.0.0

- Null safety version.

## 2.3.2

- Fixed a memory leak issue.

## 2.3.1+1

- Fix health for pub.

## 2.3.1

- Fix [#46](https://github.com/OpenFlutter/flutter_oktoast/issues/46).

## 2.3.0

New feature:

- Add animation for show/hide the oktoast.

## 2.2.1

Fix:

- `textAlign` of method `showToast`.

## 2.2.0

New feature:

- handleTouch: Caller can use this property to respond to click events.

Change:

- This is a **breaking change** : The default value of dismissOtherToast of showToast is changed to better match the OKToast overall settings. .

## 2.1.9

Fix:

- `position` is not valid in `showToastWidget`.

## 2.1.8

Add param `showAnim` param for dismiss method:

- dismissAllToast
- ToastManager.dismissAll
- ToastFuture.dismiss

## 2.1.7

Add for `OKToast`:

- textAlign
- textPadding

Add for `showToast`:

- textAlign

## 2.1.6

add textPadding for showToast method.

## 2.1.5

fix `ToastPosition` offset params

## 2.1.4

add textDirection params

## 2.1.3

When ui.window's size changes, toast is moved.

## 2.1.2

Now, `OKToast` add a params `dismissOtherOnShow` to dismiss other toast.

## 2.1.1

use manager to manage ToastFuture

add a method `dismissAllToast` to dismiss all toast.

add a param with showToast to dismiss other toast.

## 2.1.0

add new params to helper user listen toast dismiss

showToast and showToastWidget will return `ToastFuture` the toastFuture call `dismiss()` to dismiss toast.

## 2.0.1

oktoast support page level.

you can use oktoast wrap your Scaffold.

## 2.0.0

use BuildContext cache. then `showToast` not need context, you can use the `showToast` in anywhere.

## 1.0.4

support flutter sdk 0.10 ,fix bug

## 1.0.3

update the textAlign : TextAlign.center  
update the overflow: TextOverflow.ellipsis,

## 1.0.2

fix `defaultPosition` is invalid.

## [1.0.1+2]

update readme

## [1.0.1+1]

update readme

## [1.0.1] rename library name

now libray is oktoast

## [1.0.0] first release version

stable version

## [0.0.3] add the toast position

the toast can set position

1. showToast
2. defaultPosition

## [0.0.2] - add toast style

add the textStyle
radius
backgroundColor

## [0.0.1] - first version

init commit


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:oktoast/oktoast.dart';

import 'miui10_anim.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return buildApp();
  }

  // 2-A: Wrap MaterialApp with OKToast.
  Widget buildWrapperApp() {
    return OKToast(
      // 2-A: wrap your app with OKToast
      child: MaterialApp(
        title: 'Demo for OKToast',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: const MyHomePage(),
      ),
    );
  }

  // 2-B: Or wrap child of the builder method.
  Widget buildApp() {
    return MaterialApp(
      home: const MyHomePage(),
      builder: (_, Widget? child) => OKToast(
        textStyle: const TextStyle(fontSize: 19.0, color: Colors.white),
        backgroundColor: Colors.grey,
        animationCurve: Curves.easeIn,
        animationBuilder: const Miui10AnimBuilder().call,
        animationDuration: const Duration(milliseconds: 200),
        duration: const Duration(seconds: 3),
        child: child!,
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    _counter++;
    setState(() {});
  }

  void _showToast() {
    // 3-A use showToast method
    showToast(
      '$_counter',
      position: ToastPosition.bottom,
      backgroundColor: Colors.black.withOpacity(0.8),
      radius: 13.0,
      textStyle: const TextStyle(fontSize: 18.0, color: Colors.white),
      animationBuilder: const Miui10AnimBuilder().call,
    );

    showToast(
      '$_counter',
      duration: const Duration(milliseconds: 3500),
      position: ToastPosition.top,
      backgroundColor: Colors.black.withOpacity(0.8),
      radius: 3.0,
      textStyle: const TextStyle(fontSize: 30.0, color: Colors.white),
    );

    // 3-B use showToastWidget method
    final Widget widget = Center(
      child: ClipRRect(
        borderRadius: BorderRadius.circular(30.0),
        child: Container(
          width: 40.0,
          height: 40.0,
          color: Colors.grey.withOpacity(0.3),
          child: const Icon(
            Icons.add,
            size: 30.0,
            color: Colors.green,
          ),
        ),
      ),
    );
    final ToastFuture toastFuture = showToastWidget(
      widget,
      duration: const Duration(seconds: 3),
      onDismiss: () {
        // The method will be called on toast dismiss.
        debugPrint('Toast has been dismissed.');
      },
    );

    // can use future
    Future<void>.delayed(const Duration(seconds: 2), () {
      toastFuture.dismiss(); // dismiss
    });
  }

  ToastFuture? _persistToast;

  void _showPersistToast() {
    _persistToast = showToastWidget(
      Center(
        child: ElevatedButton(
          onPressed: () => _persistToast?.dismiss(),
          child: const Text('Click this button to dismiss'),
        ),
      ),
      duration: Duration.zero,
      handleTouch: true,
    );
  }

  void _dismissToastSynchronously() {
    final ToastFuture toast = showToast('Synchronously dismiss');
    toast.dismiss();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Example for OKToast')),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Center(
              child: Text('You have pushed the button this many times:'),
            ),
            Center(
              child: Text(
                '$_counter',
                style: const TextStyle(
                  fontSize: 20.0,
                  fontWeight: FontWeight.w600,
                ),
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: ElevatedButton(
                onPressed: _showToast,
                child: const Text('Show toast'),
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: ElevatedButton(
                onPressed: _showPersistToast,
                child: const Text('Show a persist toast'),
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: ElevatedButton(
                child: const Text('Toast during pushing to a new page'),
                onPressed: () {
                  _showToast();
                  Navigator.push(
                    context,
                    MaterialPageRoute<void>(builder: (_) => const MyHomePage()),
                  );
                },
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: ElevatedButton(
                onPressed: _dismissToastSynchronously,
                child: const Text('Synchronously dismiss toast'),
              ),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Add number',
        child: const Icon(Icons.add),
      ),
    );
  }
}
